cin_validator/rules/cin2022_23/rule_8915.py

```diff
@@ -0,0 +1,214 @@
+Module: Child protection plans
+Rule details: If <PersonDeathDate> (N00108) is present, then <CPPstartDate> (N00105) must be on or before <PersonDeathDate> (N00108)
+Rule message: Child Protection Plan shown as starting after the child’s Date of Death
+
+"""
+from typing import Mapping
+
+import pandas as pd
+
+from cin_validator.rule_engine import rule_definition, CINTable, RuleContext
+from cin_validator.rule_engine import IssueLocator
+from cin_validator.test_engine import run_rule
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+ChildProtectionPlans = CINTable.ChildProtectionPlans
+CPPstartDate = ChildProtectionPlans.CPPstartDate
+CPP_LAID = ChildProtectionPlans.LAchildID
+
+ChildIdentifiers = CINTable.ChildIdentifiers
+PersonDeathDate = ChildIdentifiers.PersonDeathDate
+CI_LAID = ChildIdentifiers.LAchildID
+
+
+# define characteristics of rule
+@rule_definition(
+    code=8915,
+    module=CINTable.ChildProtectionPlans,
+    message="Child Protection Plan shown as starting after the child’s Date of Death",
+    affected_fields=[
+        CPPstartDate,
+        PersonDeathDate,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    df_CPP = data_container[ChildProtectionPlans].copy()
+    df_CI = data_container[ChildIdentifiers].copy()
+
+    df_CPP.index.name = "ROW_ID"
+    df_CI.index.name = "ROW_ID"
+
+    df_CPP.reset_index(inplace=True)
+    df_CI.reset_index(inplace=True)
+
+    # Remove rows with no death date
+
+    df_CI = df_CI[df_CI[PersonDeathDate].notna()]
+
+    # <CPPstartDate> (N00105) must be on or before <PersonDeathDate>
+
+    # Join 2 tables together
+
+    df = df_CPP.merge(
+        df_CI,
+        left_on=["LAchildID"],
+        right_on=["LAchildID"],
+        how="left",
+        suffixes=("_CPP", "_CI"),
+    )
+
+    # Return those where dates don't align
+    df = df[df[CPPstartDate] > df[PersonDeathDate]].reset_index()
+
+    df["ERROR_ID"] = tuple(zip(df["LAchildID"], df[CPPstartDate], df[PersonDeathDate]))
+
+    df_CPP_issues = (
+        df_CPP.merge(df, left_on="ROW_ID", right_on="ROW_ID_CPP")
+        .groupby("ERROR_ID")["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+    print(df_CPP_issues)
+
+    df_CI_issues = (
+        df_CI.merge(df, left_on="ROW_ID", right_on="ROW_ID_CI")
+        .groupby("ERROR_ID")["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+
+    rule_context.push_type_2(
+        table=ChildProtectionPlans, columns=[CPPstartDate], row_df=df_CPP_issues
+    )
+    rule_context.push_type_2(
+        table=ChildIdentifiers, columns=[PersonDeathDate], row_df=df_CI_issues
+    )
+
+
+def test_validate():
+    # Create some sample data such that some values pass the validation and some fail.
+    sample_CPP = pd.DataFrame(
+        [
+            {
+                "LAchildID": "child1",
+                "CPPstartDate": "26/05/2000",  # Pass, same date
```

Comment:
I've just added this to make sure that a row doesn't fail if someone has a death date and no CP plan